src/nccl_ofi_idpool.cpp

```diff
@@ -122,100 +78,45 @@ int nccl_ofi_idpool_allocate_id(nccl_ofi_idpool_t *idpool)
 		}
 	}
 
-	nccl_net_ofi_mutex_unlock(&idpool->lock);
-
-	if (!found || id >= idpool->size) {
-		NCCL_OFI_WARN("No IDs available (max: %lu)", idpool->size);
-		return -ENOMEM;
+	if (!found || id >= size) {
+		NCCL_OFI_WARN("No IDs available (max: %lu)", size);
+		return FI_KEY_NOTAVAIL;
 	}
 
 	return id;
 }
 
-/*
- * @brief	Free an ID from the pool
- *
- * Return input ID into the pool.
- *
- * This operation is locked by the ID pool's internal lock.
- *
- * @param	idpool
- *		The ID pool
- * @param	id
- *		The ID to release (zero-based)
- * @return	0 on success
- *		non-zero on error
- */
-int nccl_ofi_idpool_free_id(nccl_ofi_idpool_t *idpool, size_t id)
+
+void nccl_ofi_idpool_t::free_id(size_t id)
 {
-	assert(NULL != idpool);
+	std::lock_guard<std::mutex> l(lock);
 
-	if (0 == idpool->size) {
+	if (0 == size) {
 		NCCL_OFI_WARN("Cannot free an ID from a 0-sized pool");
-		return -EINVAL;
-	}
-
-	if (OFI_UNLIKELY(NULL == idpool->ids)) {
-		NCCL_OFI_WARN("Invalid call to nccl_ofi_free_id with uninitialized pool");
-		return -EINVAL;
+		throw std::runtime_error("");
 	}
 
-	if (OFI_UNLIKELY(id >= idpool->size)) {
-		NCCL_OFI_WARN("ID value %lu out of range (max: %lu)", id, idpool->size);
-		return -EINVAL;
+	if (OFI_UNLIKELY(id >= size)) {
+		NCCL_OFI_WARN("ID value %lu out of range (max: %lu)", id, size);
+		throw std::runtime_error("");
```

Comment:
we can be a bit more helpful than an empty string :).

Reply:
Added more descriptive exception messages in latest force push.